File: 21-Memento/src/go/history.go
package main

// History is the Caretaker — stores mementos on a stack for undo. It never
// reads the state inside a memento; it only holds and returns them.
type History struct {
	mementos []*EditorMemento
}

func (h *History) Push(memento *EditorMemento) {
	h.mementos = append(h.mementos, memento)
}

func (h *History) Pop() *EditorMemento {
	last := h.mementos[len(h.mementos)-1]
	h.mementos = h.mementos[:len(h.mementos)-1]
	return last
}

func (h *History) IsEmpty() bool {
	return len(h.mementos) == 0
}

File: 21-Memento/src/rust/history.rs
use crate::editor_memento::EditorMemento;

/// Caretaker — stores mementos on a stack for undo. It never reads the state
/// inside a memento; it only holds and returns them.
pub struct History {
    mementos: Vec<EditorMemento>,
}

impl History {
    pub fn new() -> Self {
        History { mementos: Vec::new() }
    }

    pub fn push(&mut self, memento: EditorMemento) {
        self.mementos.push(memento);
    }

    pub fn pop(&mut self) -> Option<EditorMemento> {
        self.mementos.pop()
    }

    pub fn is_empty(&self) -> bool {
        self.mementos.is_empty()
    }
}
